dotnet/src/Connectors/Connectors.Memory.AzureAISearch/AzureAISearchMemoryRecordService.cs

```diff
@@ -0,0 +1,391 @@
+    /// <returns>The collection name to use.</returns>
+    private string ChooseCollectionName(string? operationCollectionName)
+    {
+        var collectionName = operationCollectionName ?? this._options.DefaultCollectionName;
+        if (collectionName is null)
+        {
+#pragma warning disable CA2208 // Instantiate argument exceptions correctly
+            throw new ArgumentException("Collection name must be provided in the operation options, since no default was provided at construction time.", "options");
+#pragma warning restore CA2208 // Instantiate argument exceptions correctly
+        }
+
+        return collectionName;
+    }
+
+    /// <summary>
+    /// Convert the public <see cref="GetRecordOptions"/> options model to the azure ai search <see cref="GetDocumentOptions"/> options model.
+    /// </summary>
+    /// <param name="options">The public options model.</param>
+    /// <returns>The azure ai search options model.</returns>
+    private GetDocumentOptions ConvertGetDocumentOptions(GetRecordOptions? options)
+    {
+        var innerOptions = new GetDocumentOptions();
+        if (options?.IncludeVectors is false)
+        {
+            innerOptions.SelectedFields.AddRange(this._nonVectorPropertyNames);
+        }
+
+        return innerOptions;
+    }
+
+    /// <summary>
+    /// Run the given operation and wrap any <see cref="RequestFailedException"/> with <see cref="HttpOperationException"/>."/>
+    /// </summary>
+    /// <typeparam name="T">The response type of the operation.</typeparam>
+    /// <param name="operation">The operation to run.</param>
+    /// <param name="collectionName">The name of the collection the operation is being run on.</param>
+    /// <param name="operationName">The type of database operation being run.</param>
+    /// <returns>The result of the operation.</returns>
+    private static async Task<T> RunOperationAsync<T>(Func<Task<T>> operation, string collectionName, string operationName)
+    {
+        try
+        {
+            return await operation.Invoke().ConfigureAwait(false);
+        }
+        catch (AggregateException ex) when (ex.InnerException is RequestFailedException innerEx)
+        {
+            var wrapperException = new MemoryServiceCommandExecutionException("Call to memory service failed.", ex);
+
+            // Using Open Telemetry standard for naming of these entries.
+            // https://opentelemetry.io/docs/specs/semconv/attributes-registry/db/
+            wrapperException.Data.Add("db.system", "AzureAISearch");
+            wrapperException.Data.Add("db.collection.name", collectionName);
+            wrapperException.Data.Add("db.operation.name", operationName);
+
+            throw wrapperException;
+        }
+        catch (RequestFailedException ex)
+        {
+            var wrapperException = new MemoryServiceCommandExecutionException("Call to memory service failed.", ex);
```

Comment:
How about `MemoryServiceOperationException`?

Reply:
I'm using the same naming as entity framework is using and I think it's nice to be using similar names for anyone already familiar with that.